LF-3134: Cleaning tasks not updating null values on task complete

SayakaOno opened the pull request:
**Description**

This PR fixes an issue where null values were not being added (e.g. product-related fields in cleaning and pest control tasks) to the DB on task completion. The major cause was the frontend not sending null values. The backend has also been updated to accept them.

Details:
- Update `cleaningTaskModel` and `pestControlTaskModel` to accept `null` for `product_quantity`.
- Add tests for task completion. (`npm run test /task.test`)
- Update `CleaningTask` and `PestControlTask` components to properly clear product info when removed.
- Update `hookFormPersistUnMount` (triggered during page transitions) not to remove `null` values for specified properties. (`propertiesToKeep` parameter is added to the `getCorrectedPayload` function)

NOTE:
- For irrigation tasks, there's currently no logic on either frontend or backend to remove irrelevant fields based on the selected method type. These fields are hidden in the UI, so I'll leave this as is.
- I plan to add tests for plant and transplant tasks later when I work on task re-completion tests. (I didn't have the energy left for that in this PR...)

Jira link: https://lite-farm.atlassian.net/browse/LF-3134

**Type of change**

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**How Has This Been Tested?**

- [x] Passes test case
- [x] UI components visually reviewed on desktop view
- [ ] UI components visually reviewed on mobile view
- [ ] Other (please explain)

**Checklist:**

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] The precommit and linting ran successfully
- [ ] I have added or updated language tags for text that's part of the UI
- [ ] I have ordered translation keys alphabetically (optional: run `pnpm i18n` to help with this)
- [x] I have added [the GNU General Public License](https://lite-farm.atlassian.net/l/cp/BT0Dd7WW) to all new files

SayakaOno commented on packages/webapp/src/components/Task/CleaningTask/index.jsx:
```diff
@@ -28,9 +28,9 @@ const PureCleaningTask = ({
     if (isCleaningAgentUsed === false && getValues('cleaning_task.product')) {
-      setValue('cleaning_task.product', null);
+      setValue('cleaning_task.product', undefined);
       setValue('cleaning_task.product_id', null);
-      setValue('cleaning_task.product_quantity', undefined);
+      setValue('cleaning_task.product_quantity', null);
```
`cleaning_task.product` should not be sent to the backend, but `cleaning_task.product_quantity: null` should.

kathyavini commented:
@SayakaOno I'm not done reviewing the test files, but can I check on one part regarding the frontend object?

I was testing a cleaning task that had a product defined, and only changed the supplier field on complete. This causes an API crash for me with `ValidationError: product_translation_key: must be string` -- do you see that as well?

The original request looked like this for me on the product part:

```js
  "cleaning_task": {
    "product": {
      "product_id": 134,
      "name": "Vinegar",
      "supplier": "Updated supplier",
      "type": "cleaning_task",
      "farm_id": "f07b2536-4182-11f0-a420-acde48001122"
    },
    "task_id": 610,
    "product_id": 134,
```

and with the change to the kept `null` values it now looks like this

```js
  "cleaning_task": {
    "product": {
      "product_id": 134,
      "name": "Vinegar",
      "product_translation_key": null,
      "supplier": "Updated supplier",
      "on_permitted_substances_list": null,
      "type": "cleaning_task",
      "farm_id": "f07b2536-4182-11f0-a420-acde48001122",
      "soil_amendment_product": null
    },
    "task_id": 610,
    "product_id": 134,
 ```
 
The crash is because `createOrPatchProduct()` is missing a `try`/`catch` (which is an easy fix), but I wans't sure if the frontend was supposed to be sending those extra `nulls` within product or not 🤔

SayakaOno commented:
@kathyavini
Thank you so much for catching that! I was expecting it might affect something...

I think sending a value only for properties that should be updated (including `null`) is the right approach.
I haven't looked into the backend yet, but if it's not expecting those properties, the request body should probably be formatted before the API call.

I'll update the model if that makes sense, but do you agree that we should format the request before the API call?
I'm wondering if updating `getCorrectedPayload` to preserve all nulls under `cleaning_task` was a bad idea.

kathyavini commented:
@SayakaOno I don't know, I think I was initially wondering if the response body DID needed formatting because of the inclusion of the `soil_amendment_product`, which isn't on the product table, but I see that's because it's added to the graph and returned from the API on the product endpoint.

Seeing that actually what we're sending matches the shape of the product API response exactly makes me feel like it's okay to just leave it exactly as it is now -- seems simpler. Functionally, it seems to work perfectly fine with just the model change on that translation key property!

(However, I think I would like error handling -- probably the nicely formatted `handleObjectionError` -- added to the `createOrPatchProduct()`! I was really surprised when a model validation crashed the backend 🤣)

SayakaOno commented:
@kathyavini Thank you for looking into it! I've updated the model and added a try catch block to `createOrPatchProduct`.

I'd appreciate a re-review, thank you! 🙏

kathyavini commented on packages/api/tests/utils/taskCompletionTestCases.js:
```diff
@@ -0,0 +1,267 @@
+    getFakeCompletionData: (initialData) => ({
+      irrigation_task: {
+        task_id: initialData.task_id,
+        irrigation_type_id: initialData.irrigation_type_id,
```
Okay in this case, the initial data IS just `null` though, right, for `irrigation_type_id` and `irrigation_type_name`? I guess `irrigation_type_name` being in the irrigationTaskFakeCompletionData threw me off a little about what is being tested here, since this also creates a `newIrrigationType`, same as the other test case. I don't think it's wrong, but just wanted to make sure that was intended and I'm understanding correctly.

SayakaOno replied:
I struggled a bit while writing tests for the irrigation task and probably got confused trying different values. I added `irrigation_type_name` as a workaround, but the test is passing without it...😓 
Sorry for the confusion, I’ll go through it again and clean things up. Thank you for catching the details!

SayakaOno commented:
Thank you Joyce!
The debugging behaviour you described feels like what happens when there's a defect in the test. I'm not sure if what I'm seeing is exactly the same, but the soil amendment task test might have an issue...
It would be great if you could show me what you saw when we have time in tech daily!
